### PR TITLE
`nn.aggr.Set2Set`

### PR DESCRIPTION
**Pull request by rusty1s**

### CODE REVIEW

File: torch_geometric/nn/aggr/base.py [hunk 1]
```diff
@@ -58,8 +48,43 @@ def __call__(self, x: Tensor, index: Optional[Tensor] = None, *,
             raise ValueError(f"Encountered mismatch between 'dim_size' (got "
                              f"'{dim_size}') and 'ptr' (got '{ptr.size(0)}')")
 
+        if index is None and ptr is None:
```
Comment (Padarn) [inline torch_geometric/nn/aggr/base.py:51]: This is much cleaner than handling index possibly being `None` everywhere 👍

File: torch_geometric/nn/aggr/base.py [hunk 2]
```diff
@@ -58,8 +48,43 @@ def __call__(self, x: Tensor, index: Optional[Tensor] = None, *,
...
+            index = x.new_zeros(x.size(dim), dtype=torch.long)
+
+        if dim_size is None and ptr is not None:
+            dim_size = ptr.numel() - 1
+        elif dim_size is None:
+            dim_size = int(index.max()) + 1 if index.numel() > 0 else 0
+
```
Comment (Padarn) [inline torch_geometric/nn/aggr/base.py:58]:
```suggestion
elif index is not None:
    assert dim_size >= index.numel() + 1
else: 
   assert dim_size >= ptr.numel() -1
```
what do you think?

File: torch_geometric/nn/aggr/base.py [hunk 3]
```diff
@@ -71,8 +96,15 @@ def reduce(self, x: Tensor, index: Optional[Tensor] = None,
...
         return scatter(x, index, dim=dim, dim_size=dim_size, reduce=reduce)
 
-    def __repr__(self) -> str:
-        return f'{self.__class__.__name__}()'
+    def to_dense_batch(self, x: Tensor, index: Optional[Tensor] = None,
+                       ptr: Optional[Tensor] = None,
+                       dim_size: Optional[int] = None,
+                       dim: int = -2) -> Tuple[Tensor, Tensor]:
+
+        self.assert_index_present(index)  # TODO
```
Comment (Padarn) [inline torch_geometric/nn/aggr/base.py:104]:
```suggestion
        self.assert_index_present(index)  # TODO: support non-sorted index
```
Unclear what the TODO was for me exactly (not sure I'm right here, suggest a better message if mine is not accurate)

File: torch_geometric/nn/aggr/set2set.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.lstm.reset_parameters()
+
+    def forward(self, x: Tensor, index: Optional[Tensor] = None, *,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
+        self.assert_index_present(index)  # TODO
```
Comment (Padarn) [inline torch_geometric/nn/aggr/set2set.py:47]: Maybe we should comment in the `assert_index_present` to make it clear we want to eventually want `ptr` to be supported everywhere?

File: torch_geometric/nn/aggr/base.py (hunk 2 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/base.py:58]: done :)

File: torch_geometric/nn/aggr/base.py (hunk 3 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/base.py:110]: done.

File: torch_geometric/nn/aggr/set2set.py (hunk 4 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/set2set.py]: done.

**Review by Padarn — APPROVED** [review]

**Pull request merged**